mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxACL.java

```diff
@@ -19,235 +19,710 @@
-         * 
-         * @see MailboxACL#DEFAULT_GROUP_MARKER
-         * @see MailboxACL#DEFAULT_NEGATIVE_MARKER
-         * 
-         * @return serialized form as a {@link String}
-         */
-        String serialize();
-    }
+    /**
+     * Marks negative when (de)serializing {@link MailboxACLEntryKey}s.
+     *
+     * @see MailboxACLEntryKey#serialize()
+     */
+    public static final char DEFAULT_NEGATIVE_MARKER = '-';
+
+    /**
+     * SETACL third argument prefix
+     */
+    public static final char REMOVE_RIGHTS_MARKER = '-';
 
     /**
      * Single right applicable to a mailbox.
      */
-    interface MailboxACLRight {
+    public enum Right {
+        Administer('a'), // (perform SETACL/DELETEACL/GETACL/LISTRIGHTS)
+        PerformExpunge('e'), //perform EXPUNGE and expunge as a part of CLOSE
+        Insert('i'), //insert (perform APPEND, COPY into mailbox)
+        /*
+        * create mailboxes (CREATE new sub-mailboxes in any
+        * implementation-defined hierarchy, parent mailbox for the new mailbox
+        * name in RENAME)
+        * */
+        CreateMailbox('k'),
+        Lookup('l'), //lookup (mailbox is visible to LIST/LSUB commands, SUBSCRIBE mailbox)
+        Post('p'), //post (send mail to submission address for mailbox, not enforced by IMAP4 itself)
+        Read('r'), //read (SELECT the mailbox, perform STATUS)
+        /**
+         * keep seen/unseen information across sessions (set or clear \SEEN
+         * flag via STORE, also set \SEEN during APPEND/COPY/ FETCH BODY[...])
+         */
+        WriteSeenFlag('s'),
+        DeleteMessages('t'), //delete messages (set or clear \DELETED flag via STORE, set \DELETED flag during APPEND/COPY)
+        /**
+         * write (set or clear flags other than \SEEN and \DELETED via
+         * STORE, also set them during APPEND/COPY)
+         */
+        Write('w'),
+        DeleteMailbox('x'); //delete mailbox (DELETE mailbox, old mailbox name in RENAME)
+
+        private final char rightCharacter;
+
+        Right(char rightCharacter) {
+            this.rightCharacter = rightCharacter;
+        }
+
         /**
          * Returns the char representation of this right.
-         * 
+         *
          * @return char representation of this right
          */
-        char asCharacter();
+        public char asCharacter() {
+            return rightCharacter;
+        }
+
+        public static final EnumSet<Right> allRights = EnumSet.allOf(Right.class);
+
+        public static Right forChar(char c) throws UnsupportedRightException {
+            return Right.allRights
+                .stream()
+                .filter(r -> r.asCharacter() == c)
+                .findFirst()
+                .orElseThrow(() -> new UnsupportedRightException(c));
+        }
     }
 
     /**
-     * Iterable set of {@link MailboxACLRight}s.
-     * 
+     * Iterable set of {@link MailboxACL.MailboxACLRight}s.
```

Comment:
MailboxACL.Right? is it yet a set?

Reply:
Pfffffff outdated javadoc. Should we also rewrite this? :-(